source/extensions/filters/network/zookeeper_proxy/zookeeper_decoder.cc

```diff
@@ -0,0 +1,344 @@
+  ENVOY_LOG(trace, "zookeeper_proxy: decoding {} bytes at offset {}", data.length(), offset);
+
+  // Check message length.
+  int32_t len = BufferHelper::peekInt32(data, offset);
+  checkLength(len, 8);
+
+  // "Special" requests.
+  int32_t xid = BufferHelper::peekInt32(data, offset);
+  switch (xid) {
+  case enumToIntSigned(XidCodes::CONNECT_XID):
+    parseConnect(data, offset, len);
+    return;
+  case enumToIntSigned(XidCodes::PING_XID):
+    // Skip opcode.
+    offset += 4;
+    callbacks_.onPing();
+    return;
+  case enumToIntSigned(XidCodes::AUTH_XID):
+    parseAuthRequest(data, offset, len);
+    return;
+  }
+
+  // "Regular" requests.
+  int32_t opcode = BufferHelper::peekInt32(data, offset);
+  switch (opcode) {
+  case enumToInt(OpCodes::GETDATA):
+    parseGetDataRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::CREATE):
+    parseCreateRequest(data, offset, len, false);
+    break;
+  case enumToInt(OpCodes::CREATE2):
+    parseCreateRequest(data, offset, len, true);
+    break;
+  case enumToInt(OpCodes::SETDATA):
+    parseSetRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::GETCHILDREN):
+    parseGetChildrenRequest(data, offset, len, false);
+    break;
+  case enumToInt(OpCodes::GETCHILDREN2):
+    parseGetChildrenRequest(data, offset, len, true);
+    break;
+  case enumToInt(OpCodes::DELETE):
+    parseDeleteRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::EXISTS):
+    parseExistsRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::GETACL):
+    parseGetAclRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::SETACL):
+    parseSetAclRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::SYNC):
+    parseSyncRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::CHECK):
+    parseCheckRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::MULTI):
+    parseMultiRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::RECONFIG):
+    parseReconfigRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::SETWATCHES):
+    parseSetWatchesRequest(data, offset, len);
+    break;
+  case enumToInt(OpCodes::CHECKWATCHES):
+    parseXWatchesRequest(data, offset, len, OpCodes::CHECKWATCHES);
+    break;
+  case enumToInt(OpCodes::REMOVEWATCHES):
+    parseXWatchesRequest(data, offset, len, OpCodes::REMOVEWATCHES);
+    break;
+  case enumToIntSigned(OpCodes::CLOSE):
+    callbacks_.onCloseRequest();
+    break;
+  default:
+    throw EnvoyException(fmt::format("Unknown opcode: {}", opcode));
+  }
+}
+
+void DecoderImpl::checkLength(const int32_t len, const int32_t minlen) const {
+  if (len < minlen) {
+    throw EnvoyException("Package is too small");
+  }
+}
+
+void DecoderImpl::parseConnect(Buffer::Instance& data, uint64_t& offset, uint32_t len) {
+  checkLength(len, 28);
+
+  // Read password - skip zxid, timeout, and session id.
+  offset += 20;
```

Comment:
I think this would be easier to follow if you had named constants with the size of various elements, so something like `offset += (ZXID_LENGTH + INT_LENGTH + INT_LENGTH)`